code/addons/test/src/node/test-manager.ts

```diff
@@ -37,42 +39,69 @@ export class TestManager {
     this.vitestManager.startVitest().then(() => options.onReady?.());
   }
 
-  async restartVitest(watchMode = false) {
+  async restartVitest({ watchMode, coverage }: { watchMode: boolean; coverage: boolean }) {
     await this.vitestManager.vitest?.runningPromise;
     await this.vitestManager.closeVitest();
-    await this.vitestManager.startVitest(watchMode);
+    await this.vitestManager.startVitest({ watchMode, coverage });
   }
 
-  async handleConfigChange(payload: TestingModuleConfigChangePayload) {
-    // TODO do something with the config
-    const config = payload.config;
+  async handleConfigChange(
+    payload: TestingModuleConfigChangePayload<any, { coverage: boolean; a11y: boolean }>
+  ) {
+    if (payload.providerId !== TEST_PROVIDER_ID) {
+      return;
+    }
+    if (this.coverage !== payload.config.coverage) {
+      try {
+        this.coverage = payload.config.coverage;
+        await this.restartVitest({ watchMode: this.watchMode, coverage: this.coverage });
+      } catch (e) {
+        this.reportFatalError('Failed to change coverage mode', e);
+      }
+    }
   }
 
   async handleWatchModeRequest(payload: TestingModuleWatchModeRequestPayload) {
-    // TODO do something with the config
-    const config = payload.config;
-
     try {
       if (payload.providerId !== TEST_PROVIDER_ID) {
         return;
       }
 
       if (this.watchMode !== payload.watchMode) {
         this.watchMode = payload.watchMode;
-        await this.restartVitest(this.watchMode);
+        await this.restartVitest({ watchMode: this.watchMode, coverage: false });
       }
     } catch (e) {
       this.reportFatalError('Failed to change watch mode', e);
     }
   }
 
-  async handleRunRequest(payload: TestingModuleRunRequestPayload) {
+  async handleRunRequest(payload: TestingModuleRunRequestPayload<Config>) {
     try {
       if (payload.providerId !== TEST_PROVIDER_ID) {
         return;
       }
 
+      const allTestsRun = (payload.storyIds ?? []).length === 0;
+      if (this.coverage) {
```

Comment:
Should we ensure the `payload.config.coverage` is in sync with `this.coverage`?

Reply:
hmm I'm not sure what the resolution is. they can definetly get out of sync, if the server is restarted, the UI will have coverage enabled while the server won't. right now, everywhere, "the server is always right", and so if the UI is out of sync, it's going to be wrong. I'm hesitant to try and fix that right here with coverage, because it will only be fixed on place out of multiple. At least right now we're consistent with the sync issue.

I'd rather fix the sync issue for good, on a higher leve.

Reply:
I'm okay with that resolution if @ghengeveld is. (He asked me to include that data into the payload)

Reply:
I would say if there's a coverage setting included in the payload, that one should take precedence over the server state here.

Reply:
alright, I've fixed it for coverage here.